flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,789 @@
+        pl.any_horizontal([pl.col(f"{c}_mean").is_not_null() for c in test_cols])
+    )
+
+    with pl.Config(tbl_cols=-1):
+        print(df_bin)
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+
+    # Get the names of all the covariance and num_points columns
+    cov_cols = [f"cov_{t1}_{t2}" for t1, t2 in product(test_cols, test_cols)]
+    n_cols = [f"count_{t1}_{t2}" for t1, t2 in product(test_cols, test_cols)]
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+
+    # Compute the farm power
+    df_bin = df_bin.with_columns(pow_farm=pl.sum_horizontal([f"{c}_mean" for c in test_cols]))
+
+    # Computed the expected farm power variance as the sum of cov_cols / n_cols
+    # This is equation 4.16
+    df_bin = df_bin.with_columns(
+        pow_farm_var=pl.sum_horizontal(
+            [pl.col(cov) / pl.col(n) for cov, n in zip(cov_cols, n_cols)]
+        )
+    )
+
+    # # If any of the cov_cols are null, set pow_farm_var to null
+    # df_bin = df_bin.with_columns(
+    #     pl.when(pl.all_horizontal([pl.col(c).is_not_null() for c in cov_cols]))
+    #     .then(pl.col("pow_farm_var"))
+    #     .otherwise(None)
+    #     .alias("pow_farm_var")
+    # )
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_bin)
+
+    # Determine the weighting of the ws/wd bins
+    df_bin, df_freq_pl = add_bin_weights(df_bin, df_freq_pl, bin_cols_without_df_name, weight_by)
+
+    # Normalize the weight column over the values in df_name column and compute
+    # the weighted farm power and weighted var per bin
+    df_bin = df_bin.with_columns(weight=pl.col("weight") / pl.col("weight").sum().over("df_name"))
+
+    # Add the weighted farm power column
+    df_bin = df_bin.with_columns(weighted_farm_power=pl.col("pow_farm") * pl.col("weight"))
+    # .with_columns(weighted_power=pl.col("pow_farm") * pl.col("weight"),
+    # weighted_power_var=pl.col("pow_farm_var") * pl.col("weight")**2)
+
+    # Now compute uplifts
+    uplift_results = {}
+    for uplift_pair, uplift_name in zip(uplift_pairs, uplift_names):
+        # Subset the dataframe to the uplift_pair in consideration
+        df_sub = df_bin.filter(pl.col("df_name").is_in(uplift_pair))
+
+        # Limit the columns to bin_cols_without_df_name, df_name, weight, pow_farm, pow_farm_var
+        df_sub = df_sub.select(
+            bin_cols_without_df_name
+            + ["df_name", "weight", "weighted_farm_power", "pow_farm", "pow_farm_var"]
+        )
+
+        # Keep bin_cols_without_df_name, weight as the index, pivot df_name across pow_farm,
+        # pow_farm_var
+        df_sub = df_sub.pivot(
+            on="df_name",
+            index=bin_cols_without_df_name,  # + ["weight"],
+        )
+
+        # Assign the weight column to be the mean of weight_[uplift_pair[0]]
+        # and weight_[uplift_pair[1]]
+        df_sub = df_sub.with_columns(
+            weight=(pl.col(f"weight_{uplift_pair[0]}") + pl.col(f"weight_{uplift_pair[1]}")) / 2
+        )
```

Comment:
Are there situations where the weights could be different for the two uplift pair data frames?

Reply:
No, not really anyway.  I had originally for this reason put weight as index value (see comment above).  This was the bug yesterday, some how numerically these identical values can can classified as different leading to wrong splits in groups.  I think having a float, especially a computed one, as a grouping index is probably just a bad idea.  This way is a little clunky but I think much safer

Reply:
OK, interesting, good catch!